### PR TITLE
azure-json, fix deserialization bug on ManagementError subclasses

### PR DESCRIPTION
**Pull request by XiaofeiCao**
- fix https://github.com/Azure/autorest.java/issues/2826
- code
  - introduce Error Client model, now properties in `ManagementError` will properly get shadowed in child classes 
    https://github.com/Azure/autorest.java/commit/1050009be648ceb8539d706906fae3ea5459344f
  - deserialization adaption(special deserialization handling for `ManagementError` subclass: https://github.com/Azure/autorest.java/commit/e927f629b71663d0af514134e79b44243e1c3ce2
  - A minor deserialization bug fix on shadow properties:
     https://github.com/Azure/autorest.java/commit/7aa99363cb9db567cec2b21c446db1a8545d7ae0
    When properties are shadowed in child class, deserialization in `fromJson` will be something like:
    ```java
    if ("details".equals(fieldName)) { // field from parent property
        List<ManagementError> details = reader.readArray(reader1 -> Error.fromJson(reader1));
        deserializedError.details = details;
    } else if ("details.equals(fieldName)) { // field from child property
        List<Error> details = reader.readArray(reader1 -> Error.fromJson(reader1));
        deserializedError.details = details;
    }
    ```
    After the fix, we will only deserialize those with the same serialized name in child classes.

### CODE REVIEW

File: typespec-tests/src/main/java/com/cadl/armstreamstyleserialization/models/Error.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+                    return fromJson0(bufferedReader);
+                } else {
+                    bufferedReader.skipChildren();
+                }
+            }
+            return fromJson0(bufferedReader.reset());
+        });
+    }
+
+    private static Error fromJson0(JsonReader reader) throws IOException {
```
Comment (weidongxu-microsoft) [inline typespec-tests/src/main/java/com/cadl/armstreamstyleserialization/models/Error.java:163]: `fromJson0` :-)

Not picky about private name. core-management is `readManagementError`

**Review by weidongxu-microsoft — APPROVED** [review]
Please loop in Alan for review, when PR ready.

File: typespec-tests/src/main/java/com/cadl/armstreamstyleserialization/models/Error.java (hunk 1 above)
Comment (XiaofeiCao) [reply typespec-tests/src/main/java/com/cadl/armstreamstyleserialization/models/Error.java]: I used `fromJson0` for generalization. Though `readManagementError` seems good enough too. Updated.

File: fluentgen/src/main/java/com/azure/autorest/fluent/model/arm/ErrorClientModel.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                .name("info")
+                .serializedName("info")
+                .description("The additional info.")
+                .readOnly(true)
+                .wireType(ClassType.OBJECT)
+                .clientType(ClassType.OBJECT)
+                .build()
+        )).build();
+
+    public static Optional<ClientModel> getErrorClientModel(String modelName) {
```
Comment (alzimmermsft) [inline fluentgen/src/main/java/com/azure/autorest/fluent/model/arm/ErrorClientModel.java:90]: Seeing how this return is used, not a big fan of the `return Optional.empty()` only to check `optional.isPresent()` then `optional.get()` as this is equivalent to `return null` and checking `value != null`.

**Review by haolingdong-msft — APPROVED** [review]

File: fluentgen/src/main/java/com/azure/autorest/fluent/model/arm/ErrorClientModel.java (hunk 2 above)
Comment (XiaofeiCao) [reply fluentgen/src/main/java/com/azure/autorest/fluent/model/arm/ErrorClientModel.java:90]: Thanks Alan. 
Seems like I can chain the logic together with `ResourceClientModel.getResourceClientModel` like this:
```java
clientModel = ResourceClientModel.getResourceClientModel(name)
    .or(() -> ErrorClientModel.getErrorClientModel(name))
    .orElse(null);
```

Updated.

**Pull request merged**